### DISCUSSION TITLE
Tree-sitter for JSX-like Rbexy, how to HTML > Ruby > HTML?

### DISCUSSION THREAD

**Discussion by lunks** [tree-sitter/tree-sitter#814] [Q&A]
Hello folks,

I want to write a parse for [rbexy](https://github.com/patbenatar/rbexy), which is a JSX-like language but for Ruby. 

Rbexy allows you to do things like:

```
<a href={@to} {**root_props(class: link_classes)}>
  <span class={classes("text")}>{content}</span>
  {@icon && <Icon name={@icon} large class={classes("icon") }/>}
  {@badge && <span class={classes("badge")}>{@badge}</span>}
</a>
```

```
<div {**root_props(class: classes("container"), sjs: { data: { current_variant_class: classes("current") }})}>
  <div>
    {@variants.map { |variant|
      <Cache key={[variant, @selected_variant == variant, "product-card"]}>
         <Variant variant={variant} />
      </Cache>
    }}
  </div>
</div>
```

Looking at tree-sitter-embedded-template, it doesn't look too hard to make something that allows an HTMLesque component to have Ruby inside it. Problem is, in Ruby there might also be components and in this case I'm a bit lost on what to do.

There is a lexer and parser in the repo, written in Ruby that might be helpful:
https://github.com/patbenatar/rbexy/blob/master/lib/rbexy/lexer.rb
https://github.com/patbenatar/rbexy/blob/master/lib/rbexy/parser.rb

I don't want to reinvent the wheel and would be ideal to reuse Ruby grammar but I'm not sure how to proceed. I'm looking for help writing this either pointing me in the right direction or actually helping me coding it. I'm not very fluent in C so rewriting either parser or lexer from Ruby seems a bit daunting to me, so alternatives would be very helpful.

**Comment by maxbrunsfeld** [comment]
I think you might need be able to re-use the Ruby grammar for the expressions inside of curly braces, but need to write your own grammar for the outer HTML-like language.

Maybe you could *extend* the existing HTML grammar, kind of like how the C++ grammar [extends](https://github.com/tree-sitter/tree-sitter-cpp/blob/fb8250eef8b4cf1ce104806c50dc206e388b0e72/grammar.js#L9) the existing C grammar (passing in the C grammar as the first argument to `grammar`). When you do this, you are able to override existing rules like [this](https://github.com/tree-sitter/tree-sitter-cpp/blob/fb8250eef8b4cf1ce104806c50dc206e388b0e72/grammar.js#L38), where your rule function receives the `original` rule as its second parameter.

**Reply by lunks** [reply]
> I think you might need be able to re-use the Ruby grammar for the expressions inside of curly braces, but need to write your own grammar for the outer HTML-like language.
> 
> Maybe you could _extend_ the existing HTML grammar, kind of like how the C++ grammar [extends](https://github.com/tree-sitter/tree-sitter-cpp/blob/fb8250eef8b4cf1ce104806c50dc206e388b0e72/grammar.js#L9) the existing C grammar (passing in the C grammar as the first argument to `grammar`). When you do this, you are able to override existing rules like [this](https://github.com/tree-sitter/tree-sitter-cpp/blob/fb8250eef8b4cf1ce104806c50dc206e388b0e72/grammar.js#L38), where your rule function receives the `original` rule as its second parameter.

Thanks for the tip. How would I handle the HTMLish code inside Ruby? i.e. `<Icon />` in the first example, `<Cache />` in my second example.
